src/mono/wasm/runtime/binding_support.js

```diff
@@ -938,7 +1054,339 @@ var BindingSupportLib = {
+					return null;
+				}
+				var json = this.get_custom_marshaler_info (typePtr, marshalerFullName);
+				result = JSON.parse(json);
+				if (!result)
+					throw new Error (`Configured custom marshaler for ${fullName} could not be loaded: ${marshalerFullName}`);
+				this._custom_marshaler_info_cache.set (typePtr, result);
+			} else {
+				result = this._custom_marshaler_info_cache.get (typePtr);
+			}
+
+			// console.log(result);
+
+			return result;
+		},
+
+		_get_struct_unboxer_for_type: function (typePtr) {
+			if (!this._struct_unboxer_cache)
+				this._struct_unboxer_cache = new Map ();
+
+			if (!this._struct_unboxer_cache.has (typePtr)) {
+				var info = this._get_custom_marshaler_info_for_type (typePtr);
+				// HACK
+				if (!info)
+					info = {};
+				if (info.error)
+					console.error(`Error while configuring automatic converter for type ${this._get_type_name(typePtr)}: ${info.error}`);
+
+				var postFilter = info.postFilter;
+
+				// console.log ("postFilter", postFilter);
+
+				var convMethod = info.outputPtr;
+				if (!convMethod) {
+					if (info.typePtr)
+						console.error(`Automatic converter for type ${this._get_type_name(typePtr)} has no suitable ToJavaScript method`);
+					this._struct_unboxer_cache.set (typePtr, null);
+				} else {
+					var signature = "m";
+					var boundConverter = this.bind_method (
+						convMethod, 0, signature, this._get_type_name(typePtr) + "$ToJavaScript"
+					);
+
+					this._struct_unboxer_cache.set (typePtr, this._compile_post_filter (typePtr, boundConverter, postFilter));
+				}
+			}
+
+			return this._struct_unboxer_cache.get (typePtr);
+		},
+
+		_unbox_struct_rooted: function (unbox_buffer, mono_obj) {
+			var objSize = Module.HEAP32[(unbox_buffer / 4) | 0];
+			var classPtr = Module.HEAP32[((unbox_buffer / 4) | 0) + 1];
+			var dataOffset = unbox_buffer + 8;
+			if (!classPtr)
+				throw new Error("classPtr is null or undefined");
+
+			// console.log (`objSize ${objSize} classPtr ${classPtr} dataOffset ${dataOffset}`);
+
+			var typePtr = this.mono_wasm_class_get_type(classPtr);
+			var unboxer = this._get_struct_unboxer_for_type(typePtr);
+			if (!unboxer)
+				throw new Error (`No CustomJavaScriptMarshaler found for struct type ${this._get_type_name(typePtr)}`);
+
+			// FIXME: Pass a ReadOnlySpan or ReadOnlyMemory instead of a bare pointer
+			return unboxer (dataOffset);
+		},
+
+		_compile_pre_filter: function (typePtr, boundConverter, js) {
+			if (!js)
+				return boundConverter;
+			
+			var closure = { 
+				MONO: MONO,
+				BINDING: this,
+				typePtr: typePtr, 
+				// (js_obj, method, parmIdx) => value
+				boundConverter: boundConverter 
+			};
+			var body = [
+				"var filteredValue = null;",
+				// "console.log(`preFilter(${value})`);",
+				`{ filteredValue = ${js}; }`,
+				// "console.log(`preFilter === ${filteredValue}`);",
+				"var convertedResult = boundConverter (filteredValue, method, parmIdx);",
```

Comment:
What is `parmIdx` for?

Reply:
Automatic marshaler selection needs to look at the type of each method parameter in order to determine how to marshal it. That's also why the method pointer is there

Reply:
I forgot but also that three-arg signature ```(filteredValue, method, parmIdx)``` predates this PR. I'd be happy to change it if you want me to, all the existing converters use it. I think originally it was used for enum conversion because knowing how to do string->Enum also requires knowing the parameter type.